pkg/pillar/zedcloud/authen.go

```diff
@@ -379,73 +380,128 @@ func checkMimeProtoType(r *http.Response) bool {
-				return nil, err
+
+		case zcert.ZCertType_CERT_TYPE_CONTROLLER_SIGNING:
+			signKeyCnt++
+			if signKeyCnt > 1 {
+				errStr := fmt.Sprintf("received more than one signing cert")
+				log.Errorf("VerifySigningCertChain: " + errStr)
+				return errors.New(errStr)
 			}
-			certByte = cert.GetCert()
-			block, _ = pem.Decode(certByte)
-			if block == nil {
-				err := fmt.Errorf("VerifySigningCertChain: can not get cert block")
-				return nil, err
+
+		case zcert.ZCertType_CERT_TYPE_CONTROLLER_ECDH_EXCHANGE:
+			// nothing to do
+
+		default:
+			errStr := fmt.Sprintf("received cert with unknown type %d", cert.Type)
+			log.Errorf("VerifySigningCertChain: " + errStr)
+			return errors.New(errStr)
+		}
+	}
+
+	log.Debugf("VerifySigningCertChain: key count %d\n", keyCnt)
+	if signKeyCnt == 0 {
+		err := fmt.Errorf("VerifySigningCertChain: fail to acquire signing cert")
+		return err
+	}
+
+	// verify signature of certificates
+	for _, cert := range sm.GetCerts() {
+		switch cert.Type {
+		case zcert.ZCertType_CERT_TYPE_CONTROLLER_SIGNING:
+			certBytes := cert.GetCert()
+			log.Debugf("VerifySigningCertChain: got signing cert, len %d\n",
+				len(certBytes))
+			interm0 := interm
+			if err := validateServerSigningCert(ctx, interm0, certBytes); err != nil {
+				return err
 			}
-			leafcert, err = x509.ParseCertificate(block.Bytes)
-			if err != nil {
-				log.Errorf("VerifySigningCertChain: fail to parse certificate %v\n", err)
-				return nil, err
+		case zcert.ZCertType_CERT_TYPE_CONTROLLER_ECDH_EXCHANGE:
+			if checkEcdh {
+				certBytes := cert.GetCert()
+				interm0 := interm
+				if err := verifySignature(certBytes, interm0); err != nil {
+					return err
+				}
 			}
-			log.Debugf("VerifySigningCertChain: get signing cert %d, len %d\n",
-				keyCnt, len(cert.GetCert()))
-		} else if cert.Type == zcert.ZCertType_CERT_TYPE_CONTROLLER_ECDH_EXCHANGE {
-			// skip for the CONTROLLER_ECDH_EXECHANGE cert type here, not for signing purpose
-			log.Debugf("VerifySigningCertChain: ECDH_EXCHANGE type, ignore cert %d, len %d\n",
-				keyCnt, len(cert.GetCert()))
-		} else {
-			err := fmt.Errorf("VerifySigningCertChain: received cert with unknown type %d", cert.Type)
-			return nil, err
 		}
 	}
+	log.Debugf("VerifySigningCertChain: success\n")
+	return nil
+}
 
-	log.Debugf("VerifySigningCertChain: key count %d\n", keyCnt)
-	if leafcert == nil {
-		err := fmt.Errorf("VerifySigningCertChain: fail to acquire leaf cert")
-		return nil, err
+func validateServerSigningCert(ctx *ZedCloudContext, interm *x509.CertPool,
+	certBytes []byte) error {
+
+	if err := verifySignature(certBytes, interm); err != nil {
+		errStr := fmt.Sprintf("%v", err)
+		log.Errorf("VerifySigningCertChain: " + errStr)
+		return err
+	}
+
+	if err := updateServerCert(ctx, certBytes); err != nil {
+		errStr := fmt.Sprintf("%v", err)
+		log.Errorf("VerifySigningCertChain: " + errStr)
+		return err
+	}
```

Comment:
Have the callers call this explicitly instead of hiding it deep down.

Reply:
Sure